### PR TITLE
Add write_to_socket function and support loading of hive tables in read_from_socket function

### PR DESCRIPTION
**Pull request by JanusAsmussen**

### CODE REVIEW

File: spark_utils/common/functions.py [hunk 1]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
+) -> None:
+    """Writes data to socket
+
+    :param data: Dataframe to write
+    :param socket: Socket to write to
+    :param partition_by: List of column names to partition by
+    :param partition_count: Number of partitions to use
+    :param kwargs: Keyword arguments passed to spark write options
+    """
+    partition_count = partition_count or 200
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:98]:
```suggestion
```

File: spark_utils/common/functions.py [hunk 2]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
         .options(**kwargs) \
         .format(socket.data_format) \
         .load(socket.data_path)
+
+
+def write_to_socket(
+    data: DataFrame,
+    socket: JobSocket,
+    partition_by: List[str] = None,
+    partition_count: int = None,
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:87]:
```suggestion
    partition_count: Optional[int] = 100,
```

File: spark_utils/common/functions.py [hunk 3]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
         .read \
         .options(**kwargs) \
         .format(socket.data_format) \
         .load(socket.data_path)
+
+
+def write_to_socket(
+    data: DataFrame,
+    socket: JobSocket,
+    partition_by: List[str] = None,
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:86]:
```suggestion
    partition_by: Optional[List[str]] = None,
```

File: spark_utils/common/functions.py [hunk 4]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
+    partition_count: int = None,
+    **kwargs
+) -> None:
+    """Writes data to socket
+
+    :param data: Dataframe to write
+    :param socket: Socket to write to
+    :param partition_by: List of column names to partition by
+    :param partition_count: Number of partitions to use
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:95]:
```suggestion
    :param partition_count: Number of partitions to split result into.
```

File: spark_utils/common/functions.py [hunk 5]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
+    :param kwargs: Keyword arguments passed to spark write options
+    """
+    partition_count = partition_count or 200
+    partition_by = partition_by or []
+    writer = data \
+        .repartition(partition_count, *partition_by) \
+        .write \
+        .mode('overwrite') \
+        .options(**kwargs)
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:104]:
```suggestion
    prepared = data.repartition(partition_count, *partition_by) if partition_count else data
    
    prepared
        .write \
        .mode('overwrite') \
        .options(**write_options)
```

File: spark_utils/common/functions.py [hunk 6]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
         .format(socket.data_format) \
         .load(socket.data_path)
+
+
+def write_to_socket(
+    data: DataFrame,
+    socket: JobSocket,
+    partition_by: List[str] = None,
+    partition_count: int = None,
+    **kwargs
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:88]:
```suggestion
    write_options: Dict[str, str]
```
Using **kwargs in public APIs is a no-go :)

File: spark_utils/common/functions.py (hunk 4 above)
Comment (george-zubrienko) [reply spark_utils/common/functions.py:95]: and please add docstring for write_options, including links to spark docs for examples (csv, parquet, json maybe)

File: spark_utils/common/functions.py [hunk 7]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
+        .mode('overwrite') \
+        .options(**kwargs)
+
+    if socket.data_format in ['delta', 'parquet']:
+        writer \
+            .format(socket.data_format) \
+            .save(socket.data_path)
+    elif socket.data_format in ['hive']:
+        writer \
+            .format('delta') \
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:112]:
```suggestion
            .format(socket.data_format) \
```

File: spark_utils/common/functions.py [hunk 8]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
...
+        .repartition(partition_count, *partition_by) \
+        .write \
+        .mode('overwrite') \
+        .options(**kwargs)
+
+    if socket.data_format in ['delta', 'parquet']:
+        writer \
+            .format(socket.data_format) \
+            .save(socket.data_path)
+    elif socket.data_format in ['hive']:
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:110]:
```suggestion
    elif socket.data_format in ['hive_delta']:
```
We can also have parquet/other tables in hive

File: spark_utils/common/functions.py [hunk 9]
```diff
@@ -70,8 +70,46 @@ def read_from_socket(socket: JobSocket, spark_session: SparkSession, **kwargs) -
     :param kwargs: Keyword arguments passed to spark read options
     :return: Spark dataframe
     """
+    if socket.data_format == 'hive':
```
Comment (george-zubrienko) [inline spark_utils/common/functions.py:73]:
```suggestion
    if socket.data_format.startswith('hive'):
```

**Review by george-zubrienko — CHANGES_REQUESTED** [review]
Some comments

**Comment by JanusAsmussen** [comment]
create_package

**Comment by JanusAsmussen** [comment]
create_package

**Comment by JanusAsmussen** [comment]
create_package

**Comment by JanusAsmussen** [comment]
create_package

**Comment by JanusAsmussen** [comment]
create_package

**Comment by JanusAsmussen** [comment]
create_package

**Review by george-zubrienko — APPROVED** [review]

**Pull request merged**